examples/dynamic-lora-sidecar/sidecar/sidecar.py

```diff
@@ -0,0 +1,164 @@
+
+
+class ConfigWatcher(FileSystemEventHandler):
+    def __init__(self, callback):
+        """
+        Watches config
+
+        Args:
+            callback : Callback function taking no parameters and no return values parsed
+        """
+        self.callback = callback
+
+    def on_modified(self, event):
+        if not event.is_directory and event.src_path.endswith(".yaml"):
+            logging.info(
+                f"Config '{event.src_path}' modified! at '{current_time_human()}'"
+            )
+            self.callback()
+
+
+class LoraReconciler:
+    """
+    Reconciles adapters registered on vllm server with adapters listed in configmap in current state
+    """
+
+    def __init__(self):
+        self.deployment_name = ""
+        self.registered_adapters = {}
+        self.config_map_adapters = {}
+        self.load_configmap()
+        self.get_registered_adapters()
+
+    def load_configmap(self):
+        with open(CONFIG_MAP_FILE, "r") as f:
+            deployment = yaml.safe_load(f)["deployment"]
+            self.deployment_name = deployment.get("name", "")
+            lora_adapters = deployment["models"]
+            self.host, self.port = (
+                deployment.get("host") or "localhost",
+                deployment.get("port") or "8000",
+            )
+            self.config_map_adapters = {adapter["id"]:adapter for adapter in lora_adapters}
+
+    def get_registered_adapters(self):
+        """Retrieves all loaded models on server"""
+        url = f"http://{self.host}:{self.port}/v1/models"
+        try:
+            response = requests.get(url)
+            adapters = {adapter["id"]: adapter for adapter in response.json()["data"]}
+            self.registered_adapters = adapters
+        except requests.exceptions.RequestException as e:
+            logging.error(f"Error communicating with vLLM server: {e}")
+
+    def reconcile(self):
+        """Reconciles model server with current version of configmap"""
+        self.get_registered_adapters()
+        self.load_configmap()
+        for adapter_id, lora_adapter in self.config_map_adapters.items():
+            logging.info(f"Processing adapter {adapter_id}")
+            if lora_adapter.get("toRemove"):
+                e = self.unload_adapter(lora_adapter)
+                lora_adapter["timestamp"] = current_time_human()
+                lora_adapter["status"] = {
+                    "timestamp": current_time_human(),
+                    "operation": "unload",
+                    "errors": [e],
+                }
+            else:
+                e = self.load_adapter(lora_adapter)
+                lora_adapter["status"] = {
+                    "timestamp": current_time_human(),
+                    "operation": "load",
+                    "errors": [e],
+                }
+        self.update_status_config()
+
+    def update_status_config(self):
+        models = list(self.config_map_adapters.values())
+        deployment = {
```

Comment:
Mounted configmaps are readonly (https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates-removed/): 

> ReadOnlyAPIDataVolumes:
> Set [configMap](https://kubernetes.io/docs/concepts/storage/volumes/#configmap), [secret](https://kubernetes.io/docs/concepts/storage/volumes/#secret), [downwardAPI](https://kubernetes.io/docs/concepts/storage/volumes/#downwardapi) and [projected](https://kubernetes.io/docs/concepts/storage/volumes/#projected) [volumes](https://kubernetes.io/docs/concepts/storage/volumes/) to be mounted read-only.
> 
> Since Kubernetes v1.10, these volume types are always read-only and you cannot opt out.

So, we need to use a different file/in-mem as @liu-cong suggests. 

see logs:
```sh
configmap-reader-1 Traceback (most recent call last):
configmap-reader-1   File "/dynamic-lora-reconciler/sidecar.py", line 164, in <module>
configmap-reader-1     main()
configmap-reader-1   File "/dynamic-lora-reconciler/sidecar.py", line 149, in main
configmap-reader-1     reconcilerInstance.reconcile()
configmap-reader-1   File "/dynamic-lora-reconciler/sidecar.py", line 91, in reconcile
configmap-reader-1     self.update_status_config()
configmap-reader-1   File "/dynamic-lora-reconciler/sidecar.py", line 102, in update_status_config
configmap-reader-1     with open(CONFIG_MAP_FILE, "w") as f:
configmap-reader-1 OSError: [Errno 30] Read-only file system: '/config/configmap.yaml'
```

Reply:
removed, replaced with storing in mem and logging